File: legacy-python/data_structures/project.py
from collections import defaultdict


class Project:
    """Simple representation of a project managed by the DAO."""

    def __init__(
        self,
        dao,
        creator,
        title,
        description,
        funding_goal,
        duration=30,
        required_skills=None,
    ):
        self.dao = dao
        self.creator = creator
        self.title = title
        self.description = description
        self.funding_goal = funding_goal
        self.duration = duration
        self.current_funding = 0
        self.required_skills = required_skills or []
        self.work_done = defaultdict(float)
        self.status = "open"
        self.comments = []
        self.start_time = 0
        # Track if funding has been locked in the treasury
        self.funding_locked = False

    def add_comment(self, member, sentiment):
        self.comments.append({"member": member, "sentiment": sentiment})

    def update_work_done(self, member, work_amount):
        self.work_done[member] += work_amount

    def receive_work(self, member, work_amount):
        """Record work contributed by a member."""
        self.update_work_done(member, work_amount)

    def total_work(self) -> float:
        return sum(self.work_done.values())

    def member_share(self, member) -> float:
        total = self.total_work()
        if total == 0:
            return 0.0
        return self.work_done.get(member, 0.0) / total

    def to_dict(self):
        return {
            "title": self.title,
            "description": self.description,
            "funding_goal": self.funding_goal,
            "duration": self.duration,
            "current_funding": self.current_funding,
            "required_skills": list(self.required_skills or []),
            "status": self.status,
            "comments": [
                {"member": getattr(c.get("member"), "unique_id", None), "sentiment": c.get("sentiment")}
                for c in self.comments
            ],
            "start_time": self.start_time,
            "funding_locked": getattr(self, "funding_locked", False),
            "creator": getattr(self.creator, "unique_id", None),
        }

    @classmethod
    def from_dict(cls, data, dao, members_by_id):
        creator = members_by_id.get(data.get("creator"))
        proj = cls(
            dao,
            creator,
            data.get("title"),
            data.get("description"),
            data.get("funding_goal", 0),
            duration=data.get("duration", 0),
            required_skills=data.get("required_skills", []),
        )
        proj.current_funding = data.get("current_funding", 0)
        proj.status = data.get("status", "open")
        proj.start_time = data.get("start_time", 0)
        proj.funding_locked = data.get("funding_locked", False)
        # Comments can’t be fully restored without member objects; skip or map IDs
        proj.comments = []
        return proj

File: lib/data-structures/project.ts
// Project data structure

import type { DAO } from './dao';

export class Project {
  dao: DAO;
  creator: string; // Creator's unique ID
  uniqueId: string;
  title: string;
  description: string;
  fundingGoal: number;
  duration: number;
  currentFunding: number = 0;
  requiredSkills: string[];
  workDone: Map<string, number> = new Map(); // member ID -> work amount
  status: string = 'open';
  comments: Array<{ member: string; sentiment: string }> = [];
  startTime: number = 0;
  fundingLocked: boolean = false;

  constructor(
    dao: DAO,
    creator: string,
    title: string,
    description: string,
    fundingGoal: number,
    duration: number = 30,
    requiredSkills: string[] = []
  ) {
    this.dao = dao;
    this.creator = creator;
    this.uniqueId = '';  // Will be set by DAO when added
    this.title = title;
    this.description = description;
    this.fundingGoal = fundingGoal;
    this.duration = duration;
    this.requiredSkills = requiredSkills;
  }

  addComment(memberId: string, sentiment: string): void {
    this.comments.push({ member: memberId, sentiment });
  }

  updateWorkDone(memberId: string, workAmount: number): void {
    const current = this.workDone.get(memberId) || 0;
    this.workDone.set(memberId, current + workAmount);
  }

  receiveWork(memberId: string, workAmount: number): void {
    this.updateWorkDone(memberId, workAmount);
  }

  totalWork(): number {
    let total = 0;
    for (const work of this.workDone.values()) {
      total += work;
    }
    return total;
  }

  memberShare(memberId: string): number {
    const total = this.totalWork();
    if (total === 0) return 0;

    const memberWork = this.workDone.get(memberId) || 0;
    return memberWork / total;
  }

  toDict(): any {
    return {
      uniqueId: this.uniqueId,
      title: this.title,
      description: this.description,
      fundingGoal: this.fundingGoal,
      duration: this.duration,
      currentFunding: this.currentFunding,
      requiredSkills: [...this.requiredSkills],
      status: this.status,
      comments: this.comments.map(c => ({
        member: c.member,
        sentiment: c.sentiment,
      })),
      startTime: this.startTime,
      fundingLocked: this.fundingLocked,
      creator: this.creator,
    };
  }

  static fromDict(data: any, dao: DAO): Project {
    const project = new Project(
      dao,
      data.creator || '',
      data.title || '',
      data.description || '',
      data.fundingGoal || 0,
      data.duration || 30,
      data.requiredSkills || []
    );

    project.uniqueId = data.uniqueId || '';
    project.currentFunding = data.currentFunding || 0;
    project.status = data.status || 'open';
    project.startTime = data.startTime || 0;
    project.fundingLocked = data.fundingLocked || false;
    project.comments = [];

    return project;
  }
}
